index.js

```diff
@@ -272,12 +673,55 @@ instance.prototype.action = function(action) {
 			cmd = '/hog/playback/release/' + opt.type;
 		break;
 
-}
+		case 'masterKey':
+
+			var arg = {
+				type: "s",
+				value: opt.action
+			};
+			cmd = '/hog/hardware/'+ opt.type +'/'+ opt.mId;
+		break;
+
+		case 'masterFader':
+
+			var arg = {
+				type: "s",
+				value: opt.level
+			};
+			cmd = '/hog/hardware/fader/' + opt.mId;
+		break;
+
+		case 'hardwareKey':
+
+			var arg = {
+				type: "s",
+				value: opt.action
+			};
+			cmd = '/hog/hardware/' + opt.type;
+		break;
+
+		case 'hKey':
+
+			var arg = {
+				type: "s",
+				value: opt.action
+			};
+			cmd = '/hog/hardware/h'+ opt.hId;
+		break;
+
+		case 'midiNote':
+
+			var arg = {
+				type: "s",
```

Comment:
Is velocity really expected as a string over osc? Not integer?

Reply:
Unclear.  I would have thought that, but the original 'Go Scene', etc. commands were programmed as string as well and reported as working (with bug that all were gos instead of go, halt, resume).  Was going to be part of my conversation with the person who asked for the additional commands to test those out.  HES is lackluster in their spec detail unfortunately.

Reply:
Later today I'll try to get Hog4 PC working on my laptop and try to flush this out.  I'm not in town, so I can't bench one over at the warehouse.

Reply:
@haakonnessjoen this took a while to decipher.  Ultimately Hog4 wants all number params as float.  Using packet capture I was finding that even though the parameters were set as string and the debug was showing this, certain parameters were getting implicitly sent as float instead of string.  Thus, this was working by accident but a couple actions weren't working correctly.  I'm not sure if that behavior is expected for some reason, but it reads like a bug in the OSC implementation.

I pulled the TouchOSC templates and compared packets.  Everything is a one-to-one match now.  Couldn't test fader level actions since those only work on a physical desk, but I'm fairly confident they're good now.